### PR TITLE
Minor cleanup for optax.projections.

### PR DESCRIPTION
**Pull request by carlosgmartin**
Minor cleanup for [`optax.projections`](https://optax.readthedocs.io/en/latest/api/projections.html). Includes the following changes:

1. Change `||` to `\|` in docstring math formulas. The latter renders the double bars used in the notation for norms correctly.
2. Inline unnecessary function calls.
3. Replace float constants with ints, where possible, to avoid unnecessary conversions to float in some cases when the input is integral.
4. Remove unnecessary `scale is None` conditional in [`projection_simplex`](https://optax.readthedocs.io/en/latest/api/projections.html#optax.projections.projection_simplex).
5. Remove unnecessary use of [`jax.lax.cond`](https://docs.jax.dev/en/latest/_autosummary/jax.lax.cond.html), which can be expensive, in [`projection_l2_ball`](https://optax.readthedocs.io/en/latest/api/projections.html#optax.projections.projection_l2_ball).
6. Shorten unnecessarily verbose code.

### CODE REVIEW

File: optax/projections/_projections.py [hunk 1]
```diff
@@ -273,26 +265,20 @@ def projection_l2_ball(tree: Any, scale: float = 1.0) -> Any:
...
   """
-  l2_norm = otu.tree_l2_norm(tree)
-  factor = scale / l2_norm
-  return jax.lax.cond(
-      l2_norm <= scale,
-      lambda tree: tree,
-      lambda tree: otu.tree_scale(factor, tree),
-      operand=tree,
-  )
+  factor = scale / otu.tree_l2_norm(tree).clip(min=scale)
```
Comment (vroulet) [inline optax/projections/_projections.py:268]: For readability,
`factor = jnp.minimum(scale / otu.tree_l2_norm(tree), 1)`

File: optax/projections/_projections.py [hunk 2]
```diff
@@ -67,38 +63,38 @@ def projection_box(tree: Any, lower: Any, upper: Any) -> Any:
...
   Args:
     tree: tree to project.
-    scale: scale of the hypercube, a scalar or a tree (default: 1.0).
+    scale: scale of the hypercube, a scalar or a tree (default: 1).
 
   Returns:
     projected tree, with the same structure as ``tree``.
   """
-  return projection_box(tree, lower=0.0, upper=scale)
+  return projection_box(tree, 0, scale)
```
Comment (vroulet) [inline optax/projections/_projections.py:91]: Previous code is more readable.

**Review by vroulet — COMMENTED** [review]
Thanks, minor comments left, then we can merge.
Did you build the docs to quickly check them too?

File: optax/projections/_projections.py (hunk 1 above)
Comment (vroulet) [reply optax/projections/_projections.py:268]: `scale / jnp.maximum(otu.tree_l2_norm(tree), 1)` is wrong. If the norm of the tree is less than scale, it should not be modified. In your case if the norm of the tree is less than 1, you always scale it, see code below.

In terms of values division by 0 is ok since the minimum of inf and 1 is 1. In terms of gradients, the real issue is in the square root during the computation of the norm of a zero vector. Below are workarounds.

```python
import jax
import jax.numpy as jnp

def euclidean_proj0(x, scale=1.):
  factor = scale / jnp.maximum(jnp.linalg.norm(x), 1)
  return factor*x

def euclidean_proj1(x, scale=1.):
  factor = jnp.minimum(scale/jnp.linalg.norm(x), 1.)
  return factor*x

def euclidean_proj2(x, scale=1.):
  squared_norm = jnp.sum(x**2)
  condition = squared_norm > 0
  valid_squared_norm = jnp.where(condition, squared_norm, 1.)
  norm = jnp.where(condition, jnp.sqrt(valid_squared_norm), 0.)
  factor = jnp.minimum(scale/norm, 1.)
  return factor*x

x =  1.5*jnp.ones(4)/2
print(f'{x=}, l2 norm={jnp.linalg.norm(x)}')  
# [0.75, 0.75, 0.75, 0.75], l2 norm=1.5
print(euclidean_proj0(x, 2.))
# [1., 1., 1., 1.]
# wrong
print(euclidean_proj1(x, 2.))
# [0.75, 0.75, 0.75, 0.75]
# correct
print(jax.jacobian(euclidean_proj1)(x, 2.))
# [[1. 0. 0. 0.]
#  [0. 1. 0. 0.]
#  [0. 0. 1. 0.]
#  [0. 0. 0. 1.]]
# correct
print(jax.jacobian(euclidean_proj1)(jnp.zeros(4), 2.))
# [[nan nan nan nan]
#  [nan nan nan nan]
#  [nan nan nan nan]
#  [nan nan nan nan]]
# incorrect
print(jax.jacobian(euclidean_proj2)(jnp.zeros(4), 2.))
# [[1. 0. 0. 0.]
#  [0. 1. 0. 0.]
#  [0. 0. 1. 0.]
#  [0. 0. 0. 1.]]
# correct
```

Comment (carlosgmartin) [reply optax/projections/_projections.py:268]: @vroulet Oops, I meant `scale / jnp.maximum(otu.tree_l2_norm(tree), scale)` (I mentally work with `scale=1`).

Comment (vroulet) [reply optax/projections/_projections.py:268]: That works too. Just make sure to avoid nans at 0 in the jacobian. See below

```python
import jax
import jax.numpy as jnp

def euclidean_proj0(x, scale=1.):
  factor = scale / jnp.maximum(jnp.linalg.norm(x), scale)
  return factor*x

def euclidean_proj1(x, scale=1.):
  squared_norm = jnp.sum(x**2)
  condition = squared_norm > 0
  valid_squared_norm = jnp.where(condition, squared_norm, 1.)
  norm = jnp.where(condition, jnp.sqrt(valid_squared_norm), 0.)
  factor = scale / jnp.maximum(norm, scale)
  return factor*x 

x =  1.5*jnp.ones(4)/2
print(f'{x=}, l2 norm={jnp.linalg.norm(x)}')  
# [0.75, 0.75, 0.75, 0.75], l2 norm=1.5
print(euclidean_proj0(x, 2.))
# [0.75, 0.75, 0.75, 0.75]
# correct
print(jax.jacobian(euclidean_proj0)(x, 2.))
# [[1. 0. 0. 0.]
#  [0. 1. 0. 0.]
#  [0. 0. 1. 0.]
#  [0. 0. 0. 1.]]
# correct
print(jax.jacobian(euclidean_proj0)(jnp.zeros(4), 2.))
# [[nan nan nan nan]
#  [nan nan nan nan]
#  [nan nan nan nan]
#  [nan nan nan nan]]
# incorrect
print(jax.jacobian(euclidean_proj1)(jnp.zeros(4), 2.))
# [[1. 0. 0. 0.]
#  [0. 1. 0. 0.]
#  [0. 0. 1. 0.]
#  [0. 0. 0. 1.]]
# correct
```

Comment (carlosgmartin) [reply optax/projections/_projections.py]: @vroulet Fixed, and also added tests `test_projection_l2_ball_grad_at_zero` and `test_projection_l1_ball_grad_at_zero` to `_projections_test.py` to test for this.

**Comment by carlosgmartin** [comment]
@vroulet Let me know if you'd like me to make any other changes.

**Comment by rdyro** [comment]
Can you rebase please?

**Comment by carlosgmartin** [comment]
@rdyro Done.

File: optax/projections/_projections_test.py [hunk 3]
```diff
@@ -230,5 +230,13 @@ def test_projection_ball(self, data_key, norm):
       np.testing.assert_almost_equal(norm_fun(p), small_radius, decimal=4)
 
 
+def test_projection_l2_ball_grad_at_zero():
```
Comment (rdyro) [inline optax/projections/_projections_test.py]: I think this is improperly indented?

Comment (carlosgmartin) [reply optax/projections/_projections_test.py]: @rdyro Can you clarify what you mean? I'm not seeing any linting errors.

Comment (carlosgmartin) [reply optax/projections/_projections_test.py]: Oh, do you mean these functions should be attached to the `ProjectionsTest` class?

Comment (rdyro) [reply optax/projections/_projections_test.py]: Absltest might not pick up the tests unless they're attached to a class, I think pytest can, but I'm not sure about our internal testing.

Comment (carlosgmartin) [reply optax/projections/_projections_test.py]: Updated.
